polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -55,18 +55,64 @@
 //!    older by a predefined parameter. This process is run very rarely (say, once a day). Once the
 //!    artifact is expired it is removed from disk eagerly atomically.
 
-use crate::host::PrepareResultSender;
+use crate::{host::PrecheckResultSender, LOG_TARGET};
 use always_assert::always;
-use polkadot_node_core_pvf_common::{error::PrepareError, prepare::PrepareStats, pvf::PvfPrepData};
+use polkadot_core_primitives::Hash;
+use polkadot_node_core_pvf_common::{
+	error::PrepareError, prepare::PrepareStats, pvf::PvfPrepData, RUNTIME_VERSION,
+};
 use polkadot_node_primitives::NODE_VERSION;
 use polkadot_parachain_primitives::primitives::ValidationCodeHash;
 use polkadot_primitives::ExecutorParamsHash;
 use std::{
 	collections::HashMap,
 	path::{Path, PathBuf},
+	str::FromStr as _,
 	time::{Duration, SystemTime},
 };
 
+// A workaround for defining a `const` that is a concatenation of other constants.
+macro_rules! concat_const {
+    ($($arg:expr),*) => {{
+        // ensure inputs to be strings
+        $(const _: &str = $arg;)*
+
+        const LEN: usize = 0 $(+ $arg.len())*;
+
+        // concatenate strings as byte slices
+        const CAT: [u8; LEN] = {
+            let mut cat = [0u8; LEN];
+            let mut _offset = 0;
+
+            $({
+                const BYTES: &[u8] = $arg.as_bytes();
+
+                let mut i = 0;
+                let len = BYTES.len();
+                while i < len {
+                    cat[_offset + i] = BYTES[i];
+                    i += 1;
+                }
+                _offset += len;
+            })*
+
+            cat
+        };
+
+		// The concatenation of two string slices is guaranteed to be valid UTF8,
+		// so are the byte slices as they have the same memory layout.
+        match std::str::from_utf8(&CAT) {
+            Ok(s) => s,
+            Err(_) => panic!("Error converting bytes to str"),
+        }
+    }}
+}
+
+const RUNTIME_PREFIX: &str = "wasmtime_v";
+const NODE_PREFIX: &str = "polkadot_v";
+const ARTIFACT_PREFIX: &str =
+	concat_const!(RUNTIME_PREFIX, RUNTIME_VERSION, "_", NODE_PREFIX, NODE_VERSION);
```

Comment:
```suggestion
```

This is over the top. We don't need this macro, nor do we need to concat this at compile time. We are using format anyway to generate the name.

Reply:
You mean we do something like this instead?

```rs
fn artifact_prefix() -> String {
	format!("{}{}_{}{}", RUNTIME_PREFIX, RUNTIME_VERSION, NODE_PREFIX, NODE_VERSION)
}
```

Reply:
Yes

Reply:
This is the very first version. Now it still feels odd to compute the same thing over and over.. I understand this is not remotely close to hot and this isn't about performance. When I looked at how much cleaner the code became with the const/static compared to a function call, I felt the macro thing was worth it. A const is just so much easier to use than a function. When a value is static, it probably should be used as one.

Reply:
This is the very first version. Now it still feels odd to compute the same thing over and over.. I understand this is not remotely close to hot and this isn't about performance. I understand that the macro thing may feel like too much at the first glance. When I looked at how much cleaner the code became with the const/static compared to a function call, I felt it's worth it. A const is just so much easier to use than a function, as when a value is static, it probably should be used as one. Besides, isn't it nice that drop/dealloc/format calls are wiped off from the assembly? :))

Reply:
If it's a function you can at least put it behind a `OnceCell` or something. Otherwise I did a search and there are some crates that do this: `const_format`, `concat_const`, etc. But it would be sad to pull in a whole dependency for this. I wish std had this... Meh, let's just merge as-is (with the function). I'll press the green button once conflicts are resolved.

Reply:
> If it's a function you can at least put it behind a OnceCell or something.

Function, cell, lazy static, considered all that before doing the macro as it results in the simplest interface, a static str. I won't change it back. Let's get this merged. 🚀 I'm just trying to understand why is a function call better here.

Reply:
> I'm just trying to understand why is a function call better here.

The problem with the macro is that it's a lot of code for readers to understand, just to declare a constant. TBH, all the options are kind of meh, but given that the code is cold, optimizing for maintainability with a plain function seems like the right call.